File: frontend/src/api/modules/activityHeatmap.ts
import { api } from "../client";
import type { ActivityType } from "./activities";

/** One cell of the contribution grid. */
export interface HeatmapDay {
  /** ISO `YYYY-MM-DD`, UTC. */
  day: string;
  count: number;
  /** 0 = nothing happened; 1-4 are quartiles of this user's own activity. */
  level: 0 | 1 | 2 | 3 | 4;
}

export interface StreakSummary {
  /**
   * Consecutive active days ending today *or* yesterday — an unstarted day
   * does not read as a broken streak.
   */
  current_streak: number;
  longest_streak: number;
  longest_streak_start: string | null;
  longest_streak_end: string | null;
  total_activities: number;
  active_days: number;
  total_days: number;
  busiest_day: string | null;
  busiest_day_count: number;
  daily_average: number;
  active_day_average: number;
}

export interface ActivityTypeCount {
  activity_type: ActivityType | string;
  count: number;
}

export interface ActivityHeatmap {
  user_id: string;
  username: string;
  start_date: string;
  end_date: string;
  /** Every day in the window, ascending, gaps included as `count: 0`. */
  days: HeatmapDay[];
  streak: StreakSummary;
  breakdown: ActivityTypeCount[];
}

export interface HeatmapQuery {
  /** 1-366. Defaults to 365 server-side. */
  days?: number;
  /** Comma-separated ActivityType values; omit for everything. */
  activity_types?: string;
}

export const activityHeatmapApi = {
  getForUser: (username: string, query?: HeatmapQuery) =>
    api.get<ActivityHeatmap>(`/api/v1/users/${encodeURIComponent(username)}/activity-heatmap`, {
      query: query as Record<string, string | number | undefined>,
    }),

  getMine: (query?: HeatmapQuery) =>
    api.get<ActivityHeatmap>("/api/v1/users/me/activity-heatmap", {
      auth: true,
      query: query as Record<string, string | number | undefined>,
    }),
};

File: backend/app/schemas/activity_heatmap.py
from __future__ import annotations

"""
Schemas for the contribution heatmap and streak summary (#1040).
"""

import uuid
from datetime import date
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field

# A calendar day's intensity bucket. 0 means "nothing happened"; 1-4 are
# quartiles of the *user's own* activity in the requested window, so a quiet
# account still produces a readable gradient instead of a flat wall of level 1.
HeatmapLevel = int


class HeatmapDay(BaseModel):
    """One cell of the heatmap grid."""

    day: date = Field(description="UTC calendar day")
    count: int = Field(ge=0, description="Number of activities recorded on this day")
    level: HeatmapLevel = Field(ge=0, le=4, description="Intensity bucket, 0-4")

    model_config = ConfigDict(from_attributes=True)


class StreakSummary(BaseModel):
    """Aggregate streak statistics over the requested window."""

    current_streak: int = Field(
        ge=0,
        description=(
            "Consecutive active days ending today or yesterday. Yesterday counts "
            "so that an early-morning request does not report a broken streak."
        ),
    )
    longest_streak: int = Field(ge=0, description="Longest run of consecutive active days in the window")
    longest_streak_start: Optional[date] = Field(default=None, description="First day of the longest run")
    longest_streak_end: Optional[date] = Field(default=None, description="Last day of the longest run")

    total_activities: int = Field(ge=0, description="Total activities in the window")
    active_days: int = Field(ge=0, description="Number of days with at least one activity")
    total_days: int = Field(ge=1, description="Size of the window in days")

    busiest_day: Optional[date] = Field(default=None, description="Day with the most activity")
    busiest_day_count: int = Field(default=0, ge=0, description="Activity count on the busiest day")

    daily_average: float = Field(
        ge=0,
        description="Total activities divided by the window size, rounded to two decimals",
    )
    active_day_average: float = Field(
        ge=0,
        description="Total activities divided by active days only, rounded to two decimals",
    )


class ActivityTypeCount(BaseModel):
    """How the window's activity breaks down by type."""

    activity_type: str
    count: int = Field(ge=0)


class ActivityHeatmapResponse(BaseModel):
    """Full heatmap payload for one user."""

    user_id: uuid.UUID
    username: str

    start_date: date
    end_date: date

    days: list[HeatmapDay] = Field(
        description=(
            "One entry per day in the window, inclusive of both ends and in "
            "ascending order. Days with no activity are present with count 0 so "
            "the client can render the grid without filling gaps itself."
        )
    )
    streak: StreakSummary
    breakdown: list[ActivityTypeCount] = Field(
        default_factory=list,
        description="Activity counts grouped by type, descending",
    )

    model_config = ConfigDict(from_attributes=True)
